NetKet/Graph/lattice_impl.hpp

```diff
@@ -0,0 +1,430 @@
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_LATTICE_IMPL_HPP
+#define NETKET_LATTICE_IMPL_HPP
+
+#include <cmath>
+#include <iostream>
+#include <vector>
+#include "Utils/array_search.hpp"
+#include "Utils/messages.hpp"
+#include "Utils/next_variation.hpp"
+
+namespace netket {
+// Constructor
+Lattice::Lattice(std::vector<std::vector<double>> basis_vector,
+                 std::vector<int> extent, std::vector<bool> pbc,
+                 std::vector<std::vector<double>> V_atoms)
+    : extent_(std::move(extent)),
+      pbc_(std::move(pbc)),
+      basis_vectors_(std::move(basis_vector)),
+      atoms_coord_(std::move(V_atoms)) {
+  ndim_ = basis_vectors_.size();
+
+  // check if dimensions are consistents
+  for (int i = 0; i < ndim_; i++) {
+    if (basis_vectors_[i].size() != ndim_) {
+      ErrorMessage() << "Each element of basis_vectors must have ndim"
+                        "components. Dimensions are inconsistent. Exit.\n";
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  if (extent_.size() != ndim_) {
+    ErrorMessage() << "Extent must have ndim components. Dimension is "
+                      "inconsistent. Exit.\n";
+    exit(EXIT_FAILURE);
+  }
+
+  // default case is a single atom in the unit cell, located at the origin
+  if (atoms_coord_.size() == 0) {
+    atoms_coord_.resize(1, std::vector<double>(ndim_, 0));
+  }
+
+  natoms_ = atoms_coord_.size();
+
+  for (int i = 0; i < natoms_; i++) {
+    if (atoms_coord_[i].size() != ndim_) {
+      ErrorMessage() << "Each element of atoms_coord must have ndim "
+                        "components. Dimension is inconsistent. Exit.\n";
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  // default case for pbc is true
+  if (pbc_.size() == 0) {
+    pbc_.resize(ndim_, true);
+  }
+
+  if (pbc_.size() != ndim_) {
+    ErrorMessage() << "Pbc must have ndim components. Dimension is "
+                      "inconsistent. Exit.\n";
+    exit(EXIT_FAILURE);
+  }
+
+  // PATHOLOGIC 1: pbc(i) for extent(i)<=2 + extent(i)=0)
+  for (int i = 0; i < ndim_; i++) {
+    if (extent_[i] <= 2 and pbc_[i] == true) {
+      ErrorMessage() << "PBC are not allowed when extent<2. Exit.\n";
+      exit(EXIT_FAILURE);
+    }
+    if (extent_[i] == 0) {
+      ErrorMessage() << "Extent components must be >0. Exit.\n";
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  nlatticesites_ = 1;
+  for (int k = 0; k < ndim_; k++) nlatticesites_ *= extent_[k];
+
+  // PATHOLOGIC 2: nlatticesites_=0,1
+  if (nlatticesites_ <= 1) {
+    ErrorMessage()
+        << "A well-defined lattice must have at least 2 sites. Exit.\n";
+    exit(EXIT_FAILURE);

```

Comment:
The way we are handling errors is through exceptions, please do something like 
```c++
throw InvalidInputError{
          "A well-defined lattice must have at least 2 sites."};
```
then you can remove all the exit statements without problems

Reply:
Done!